File: python/scenario/voice/effects/quality.py
"""Quality-degradation effects: phone_quality, low_quality, packet_loss, echo, robotic, breaking_up."""

from __future__ import annotations

import numpy as np

from ._common import EffectFn, np_to_pcm16, pcm16_to_np, rate


def phone_quality() -> EffectFn:
    """Bandpass 300Hz-3.4kHz + amplitude compression to mimic a phone line."""

    def _apply(audio: bytes) -> bytes:
        arr = pcm16_to_np(audio).astype(np.float32)
        if len(arr) == 0:
            return audio
        # Simple bandpass via FFT — cheap and dependency-free.
        fft = np.fft.rfft(arr)
        freqs = np.fft.rfftfreq(len(arr), d=1.0 / rate())
        mask = (freqs >= 300) & (freqs <= 3400)
        fft *= mask
        filtered = np.fft.irfft(fft, n=len(arr))
        # Mild compression (saturate)
        compressed = np.tanh(filtered / 16000.0) * 16000.0
        return np_to_pcm16(compressed)

    return _apply


def low_quality(bitrate: int = 8000) -> EffectFn:
    """Downsample to ``bitrate`` Hz and back, simulating a low-bitrate codec."""

    def _apply(audio: bytes) -> bytes:
        arr = pcm16_to_np(audio)
        if len(arr) == 0 or bitrate >= rate():
            return audio
        # Downsample then upsample — introduces aliasing and quantisation noise.
        down_len = max(1, int(len(arr) * bitrate / rate()))
        down_idx = np.linspace(0, len(arr) - 1, down_len).astype(np.int64)
        down = arr[down_idx]
        up_idx = np.linspace(0, len(down) - 1, len(arr)).astype(np.int64)
        return np_to_pcm16(down[up_idx])

    return _apply


def packet_loss(probability: float = 0.05, chunk_ms: int = 20) -> EffectFn:
    """Zero out random ``chunk_ms`` windows at the given probability."""
    if not 0 <= probability <= 1:
        raise ValueError("packet_loss probability must be in [0, 1]")

    def _apply(audio: bytes) -> bytes:
        arr = pcm16_to_np(audio).copy()
        if len(arr) == 0:
            return audio
        chunk_samples = max(1, (rate() * chunk_ms) // 1000)
        rng = np.random.default_rng()
        for i in range(0, len(arr), chunk_samples):
            if rng.random() < probability:
                arr[i : i + chunk_samples] = 0
        return np_to_pcm16(arr)

    return _apply


def echo(delay_ms: int = 200, decay: float = 0.5) -> EffectFn:
    """Overlay a delayed/attenuated copy of the signal."""

    def _apply(audio: bytes) -> bytes:
        arr = pcm16_to_np(audio).astype(np.float32)
        if len(arr) == 0:
            return audio
        delay_samples = (rate() * delay_ms) // 1000
        if delay_samples >= len(arr):
            return audio
        delayed = np.zeros_like(arr)
        delayed[delay_samples:] = arr[:-delay_samples] * decay
        return np_to_pcm16(arr + delayed)

    return _apply


def robotic() -> EffectFn:
    """Crude vocoder-ish effect: ring-modulate the signal with a low-freq carrier."""

    def _apply(audio: bytes) -> bytes:
        arr = pcm16_to_np(audio).astype(np.float32)
        if len(arr) == 0:
            return audio
        t = np.arange(len(arr)) / rate()
        carrier = np.sin(2 * np.pi * 100 * t)
        return np_to_pcm16(arr * carrier)

    return _apply


def breaking_up() -> EffectFn:
    """Simulate intermittent connection: larger and more frequent dropouts than packet_loss."""

    def _apply(audio: bytes) -> bytes:
        arr = pcm16_to_np(audio).copy()
        if len(arr) == 0:
            return audio
        chunk_samples = (rate() * 100) // 1000  # 100ms windows
        rng = np.random.default_rng()
        for i in range(0, len(arr), chunk_samples):
            if rng.random() < 0.2:
                arr[i : i + chunk_samples] = 0
        return np_to_pcm16(arr)

    return _apply

File: javascript/src/voice/effects/quality.ts
/**
 * Quality-degradation effects: phoneQuality, lowQuality, packetLoss, echo, robotic, breakingUp.
 *
 * TS equivalent of `python/scenario/voice/effects/quality.py`.
 */

import FFT from "fft.js";

import { EffectFn, int16ToPcm16, linearResample, pcm16ToInt16, rate } from "./common";

/** Returns the next power of two >= n. */
function nextPow2(n: number): number {
  let p = 1;
  while (p < n) p <<= 1;
  return p;
}

// ---------------------------------------------------------------------------
// phoneQuality
// ---------------------------------------------------------------------------

/**
 * Bandpass 300Hz–3.4kHz via FFT + tanh compression to mimic a phone line.
 */
export function phoneQuality(): EffectFn {
  return function _apply(audio: Uint8Array): Uint8Array {
    const arr = pcm16ToInt16(audio);
    if (arr.length === 0) return audio;

    const n = arr.length;
    const fftSize = nextPow2(n);
    const fft = new FFT(fftSize);

    // Zero-pad real input into the interleaved complex format expected by realTransform
    const realInput = new Array<number>(fftSize).fill(0);
    for (let i = 0; i < n; i++) realInput[i] = arr[i]!;

    // forward transform → complex array (interleaved: [re0, im0, re1, im1, ...])
    const spectrum = fft.createComplexArray();
    fft.realTransform(spectrum, realInput);

    // Fill the negative frequencies (realTransform only fills [0..fftSize/2])
    // by conjugate symmetry so inverseTransform works correctly.
    fft.completeSpectrum(spectrum);

    // Bandpass mask: keep bins whose frequency falls in [300, 3400] Hz
    const nyquist = rate() / 2;
    const binCount = fftSize; // number of complex pairs in full spectrum
    for (let bin = 0; bin < binCount; bin++) {
      const freq = (bin * nyquist) / (binCount / 2);
      if (freq < 300 || freq > 3400) {
        spectrum[bin * 2] = 0;
        spectrum[bin * 2 + 1] = 0;
      }
    }

    // Inverse transform
    const reconstructed = fft.createComplexArray();
    fft.inverseTransform(reconstructed, spectrum);

    // Extract real parts, truncate to original length
    const filtered = new Float32Array(n);
    for (let i = 0; i < n; i++) {
      filtered[i] = reconstructed[i * 2]!;
    }

    // Mild tanh compression
    for (let i = 0; i < n; i++) {
      filtered[i] = Math.tanh(filtered[i]! / 16000) * 16000;
    }

    return int16ToPcm16(filtered);
  };
}

// ---------------------------------------------------------------------------
// lowQuality
// ---------------------------------------------------------------------------

/**
 * Downsample to `bitrate` Hz and back — simulates a low-bitrate codec.
 */
export function lowQuality(bitrate = 8000): EffectFn {
  return function _apply(audio: Uint8Array): Uint8Array {
    const arr = pcm16ToInt16(audio);
    if (arr.length === 0 || bitrate >= rate()) return audio;

    const downLen = Math.max(1, Math.floor((arr.length * bitrate) / rate()));
    const down = linearResample(arr, downLen);
    const up = linearResample(down, arr.length);

    return new Uint8Array(up.buffer);
  };
}

// ---------------------------------------------------------------------------
// packetLoss
// ---------------------------------------------------------------------------

/**
 * Zero out random `chunkMs`-sized windows at the given `probability`.
 */
export function packetLoss(probability = 0.05, chunkMs = 20): EffectFn {
  if (probability < 0 || probability > 1) {
    throw new Error("packetLoss probability must be in [0, 1]");
  }

  return function _apply(audio: Uint8Array): Uint8Array {
    const arr = new Int16Array(pcm16ToInt16(audio)); // copy
    if (arr.length === 0) return audio;

    const chunkSamples = Math.max(1, Math.floor((rate() * chunkMs) / 1000));
    for (let i = 0; i < arr.length; i += chunkSamples) {
      if (Math.random() < probability) {
        const end = Math.min(arr.length, i + chunkSamples);
        arr.fill(0, i, end);
      }
    }
    return new Uint8Array(arr.buffer);
  };
}

// ---------------------------------------------------------------------------
// echo
// ---------------------------------------------------------------------------

/**
 * Overlay a delayed/attenuated copy of the signal.
 */
export function echo(delayMs = 200, decay = 0.5): EffectFn {
  return function _apply(audio: Uint8Array): Uint8Array {
    const arr = pcm16ToInt16(audio);
    if (arr.length === 0) return audio;

    const delaySamples = Math.floor((rate() * delayMs) / 1000);
    if (delaySamples >= arr.length) return audio;

    const out = new Float32Array(arr.length);
    for (let i = 0; i < arr.length; i++) {
      out[i] = arr[i]!;
      if (i >= delaySamples) {
        out[i] += arr[i - delaySamples]! * decay;
      }
    }
    return int16ToPcm16(out);
  };
}

// ---------------------------------------------------------------------------
// robotic
// ---------------------------------------------------------------------------

/**
 * Crude vocoder-ish effect: ring-modulate the signal with a 100 Hz carrier.
 */
export function robotic(): EffectFn {
  return function _apply(audio: Uint8Array): Uint8Array {
    const arr = pcm16ToInt16(audio);
    if (arr.length === 0) return audio;

    const out = new Float32Array(arr.length);
    const sr = rate();
    for (let i = 0; i < arr.length; i++) {
      const t = i / sr;
      const carrier = Math.sin(2 * Math.PI * 100 * t);
      out[i] = arr[i]! * carrier;
    }
    return int16ToPcm16(out);
  };
}

// ---------------------------------------------------------------------------
// breakingUp
// ---------------------------------------------------------------------------

/**
 * Simulate intermittent connection: 100ms chunks, 20% drop probability.
 */
export function breakingUp(): EffectFn {
  return function _apply(audio: Uint8Array): Uint8Array {
    // Copy so we can zero in-place
    const arr = new Int16Array(pcm16ToInt16(audio));
    if (arr.length === 0) return audio;

    const chunkSamples = Math.floor((rate() * 100) / 1000); // 100ms windows
    for (let i = 0; i < arr.length; i += chunkSamples) {
      if (Math.random() < 0.2) {
        const end = Math.min(arr.length, i + chunkSamples);
        arr.fill(0, i, end);
      }
    }
    return new Uint8Array(arr.buffer);
  };
}
